### DISCUSSION TITLE
Stale default values in Next.js after submitting a form action

### DISCUSSION THREAD

**Discussion by ChristianIvicevic** [edmundhung/conform#999] [General]
### Describe the bug and the expected behavior

After my server action has resolved and returned a successful response the form items are still showing the previous default values (reverting the newly entered values) despite me calling `router.refresh()` and fetching the latest data in the parent page and passing it down to the client component that renders the form.

To circumvent this issue I attempted to pass `Date.now()` as a key prop to my client component to force a full rerender which technically works, but it's noticeable that the form resets shortly to the previous default values and then updates with the new values.

### Conform version

v1.2.2

### Steps to Reproduce the Bug or Issue

1. Clone this branch of my fork: https://github.com/ChristianIvicevic/conform/tree/nextjs-stale-defaults
2. Open the Next.js example and go to the stale page
3. See the current value read from the database file above the input and how it changes after submitting the form, but the field not updating accordingly

### What browsers are you seeing the problem on?

Chrome, Safari

### Screenshots or Videos

https://github.com/user-attachments/assets/44bb30c6-233e-47dd-99ff-aaa678b375d9

**Comment by bbaterdene-gh** [comment]
I managed to resolve my issue by using submission.reply() instead of returning an object. This approach ensures lastResult in the useForm hook stays updated as expected.

It might be helpful to adjust line 71 in examples/nextjs/app/actions.ts to use submission.reply().

**Comment by ChristianIvicevic** [comment]
This issue has been fairly old, but in the last week I changed over to use `submission.reply()` and unfortunately that never worked as expected either.

Chances are that the new React 19 changes which reset forms on submit do contribute to that behavior as well. Once I find some time in the next few days or weeks I will look into an updated reproduction to contribute in order to see whether it's an issue with conform or my custom helper.

**Comment by dariuscosden** [comment]
Having this issue as well, with `submission.reply()`. Any updates?

**Comment by mkazinauskas** [comment]
I found an issue in react for this:
https://github.com/facebook/react/issues/32362

**Comment by dariuscosden** [comment]
I actually solved my problem by making sure that I pass `initialValue: value ?? defaultValue` instead of just the default value. That way, if there is a value, it gets set over the default.

That was enough for my use case, so hopefully it helps!

**Comment by braginteractive** [comment]
In react router v7, to get the form value to update I am doing something like this:

```
	return data(submission.reply(), {
		headers: await createToastHeaders({
			type: 'success',
			title: 'Contact Updated',
			description: 'The contact has been updated successfully.',
		}),
	})
```
```
const [form, fields] = useForm({
		id: `edit-contact`,
		constraint: getZodConstraint(ContactUpdateSchema),
		onValidate({ formData }) {
			return parseWithZod(formData, { schema: ContactUpdateSchema })
		},
		defaultValue: contact,
		lastResult: navigation.state === 'idle' ? lastResult?.initialValue : null,
		shouldRevalidate: 'onBlur',
		onSubmit() {
			setOpen(false)
		},
	})
```

the example in the docs: https://conform.guide/integration/remix wasnt working for me with just:

` lastResult: navigation.state === 'idle' ? lastResult : null,`

**Comment by edmundhung** [comment]
I have been thinking about this kind of case lately, where you want to let users keep editing the form after a successful submission.

One option is to not reset the form and just return the submission result (i.e. `submission.reply()`), similar to what happens when there is an error. This keeps the form `live` so users can continue making changes.

If you do want to reset the form just to update the dirty flag, you can now pair the `isDirty` helper with the new `useFormData` hook. Just make sure you pass the latest defaultValue when computing the dirty state.

More details here: https://github.com/edmundhung/conform/discussions/973

---

I know this is a pretty old issue, but I do not think it is actually a bug. It is more about the timing of when the form should be reset, so I am moving this to Discussions for now.
